gcovr/gcov_parser.py

```diff
@@ -0,0 +1,994 @@
+    if line.startswith("function "):
+        match = _RE_FUNCTION_LINE.match(line)
+        if match is not None:
+            name, calls, returns, blocks = match.groups()
+            return _FunctionLine(
+                name,
+                _int_from_gcov_unit(calls),
+                _int_from_gcov_unit(returns),
+                _int_from_gcov_unit(blocks),
+            )
+
+    # SPECIALIZATION MARKER
+    #
+    # Structure: literally just lots of hyphens
+    if line.startswith("-----"):
+        return _SpecializationMarkerLine()
+
+    # SPECIALIZATION NAME
+    #
+    # Structure: a name starting in the first column, ending with a ":".
+    # It is not safe to make further assumptions about the layout of the identifier
+    # since it may be demangled.
+    #
+    # In theory, there is an ambiguity with code lines like "12345: 9:foo:",
+    # i.e. a line that ends with a label or other ":"
+    # *and* is executed extremely often.
+    # To protect against that, we'll check that the name does not start with a digit.
+    #
+    # The old parser was more robust
+    # because it would only consider specialization names
+    # on the line following a specialization marker.
+    if line.endswith(":"):
+        if len(line) > 2 and not line[0].isdigit():
+            return _SpecializationNameLine(line[:-1])
+
+    return None
+
+
+def _find_excluded_ranges(
+    lines: List[Tuple[int, str]],
+    *,
+    logger: Logger,
+    filename: str,
+    exclude_lines_by_pattern: Optional[str],
+) -> Callable[[int], bool]:
+    """
+    Scan through all lines to find line ranges covered by exclusion markers.
+
+    Example:
+    >>> lines = [(13, '// IGNORE'), (15, '// GCOV_EXCL_START'), (18, '//GCOV_EXCL_STOP')]
+    >>> exclude = _find_excluded_ranges(
+    ...     lines, logger=None, filename='test', exclude_lines_by_pattern = '.*IGNORE')
+    >>> [lineno for lineno in range(20) if exclude(lineno)]
+    [13, 15, 16, 17]
+    """
+    exclude_lines_by_pattern_regex = None
+    if exclude_lines_by_pattern:
+        exclude_lines_by_pattern_regex = re.compile(exclude_lines_by_pattern)
+
+    # possibly overlapping half-open ranges of lines that are excluded
+    exclude_line_ranges: List[Tuple[int, int]] = []
+
+    exclusion_stack = []
+    for lineno, code in lines:
+        if _EXCLUDE_LINE_FLAG in code:
+            # process the exclusion marker
+            #
+            # header is a marker name like LCOV or GCOVR
+            #
+            # START flags are added to the exlusion stack
+            # STOP flags remove a marker from the exclusion stack
+            for header, flag in _EXCLUDE_LINE_PATTERN.findall(code):
+
+                if flag == "START":
+                    exclusion_stack.append((header, lineno))
+                    continue
+
+                if flag == "STOP":
+                    if not exclusion_stack:
+                        logger.warn(
```

Comment:
Can a test be added to cover this?

Reply:
I added tests for 100% coverage of the gcov_parser module. It's a bit awkward to test the warnings because they have no consistent format, so I added an `_ExclusionRangeWarnings` class that encapsulates the exact messages.